File: php/src/Resource/NotificationSequences.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Resource;

use Fivexer\SDK\Fivexer;
use Fivexer\SDK\Internal\Json;
use Fivexer\SDK\Model\CreateNotificationSequence;
use Fivexer\SDK\Model\NotificationSequence;
use Fivexer\SDK\Model\UpdateNotificationSequence;

/** Sequences describe when to notify, relative to a task lifecycle trigger. */
final class NotificationSequences
{
    public function __construct(private readonly Fivexer $client)
    {
    }

    /** @return list<NotificationSequence> */
    public function list(): array
    {
        $data = $this->client->request('GET', '/notification-sequences') ?? [];
        /** @var list<NotificationSequence> */
        return Json::parseEach($data, 'sequences', [NotificationSequence::class, 'fromArray']);
    }

    public function create(CreateNotificationSequence $input): NotificationSequence
    {
        return NotificationSequence::fromArray(
            $this->client->request('POST', '/notification-sequences', $input->toArray()) ?? []
        );
    }

    public function get(string $sequenceId): NotificationSequence
    {
        return NotificationSequence::fromArray(
            $this->client->request('GET', '/notification-sequences/' . \rawurlencode($sequenceId)) ?? []
        );
    }

    /** Partial update — fields left unset keep their stored value. */
    public function update(string $sequenceId, UpdateNotificationSequence $update): NotificationSequence
    {
        return NotificationSequence::fromArray(
            $this->client->request(
                'PATCH',
                '/notification-sequences/' . \rawurlencode($sequenceId),
                $update->toArray()
            ) ?? []
        );
    }

    public function remove(string $sequenceId): void
    {
        $this->client->request('DELETE', '/notification-sequences/' . \rawurlencode($sequenceId));
    }
}

File: java/src/main/java/io/fivexer/sdk/NotificationSequences.java
package io.fivexer.sdk;

import io.fivexer.sdk.internal.Json;
import io.fivexer.sdk.model.CreateNotificationSequence;
import io.fivexer.sdk.model.NotificationSequence;
import io.fivexer.sdk.model.NotificationSequenceList;
import io.fivexer.sdk.model.UpdateNotificationSequence;
import java.util.Collections;
import java.util.List;

/** Sequences describe when to notify, relative to a task lifecycle trigger. */
public final class NotificationSequences {

    private final Fivexer client;

    NotificationSequences(Fivexer client) {
        this.client = client;
    }

    public List<NotificationSequence> list() {
        NotificationSequenceList envelope = client.request("GET", "/notification-sequences", null, null,
                NotificationSequenceList.class, false);
        return envelope.getSequences() == null ? Collections.emptyList() : envelope.getSequences();
    }

    public NotificationSequence create(CreateNotificationSequence input) {
        return client.request("POST", "/notification-sequences", input.toJson(), null,
                NotificationSequence.class, false);
    }

    public NotificationSequence get(String sequenceId) {
        return client.request("GET", "/notification-sequences/" + Json.enc(sequenceId), null, null,
                NotificationSequence.class, false);
    }

    /** Partial update — fields left unset keep their stored value. */
    public NotificationSequence update(String sequenceId, UpdateNotificationSequence update) {
        return client.request("PATCH", "/notification-sequences/" + Json.enc(sequenceId), update.toJson(),
                null, NotificationSequence.class, false);
    }

    public void remove(String sequenceId) {
        client.request("DELETE", "/notification-sequences/" + Json.enc(sequenceId), null, null,
                Void.class, true);
    }
}
